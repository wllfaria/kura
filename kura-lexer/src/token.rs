#[allow(clippy::module_inception)]
mod token;

mod kind;
mod operator;
mod primitive;
mod value;

pub use kind::Kind;
pub use operator::Operator;
pub use primitive::{FloatSizes, IntSizes, Primitive, UIntSizes};
pub use token::{IntoToken, Location, Token};
pub use value::Value;
