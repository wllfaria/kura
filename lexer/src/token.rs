#[derive(Debug)]
pub enum NumSize {
    F8,
    F16,
    F32,
    F64,
    I8,
    I16,
    I32,
    I64,
    U8,
    U16,
    U32,
    U64,
    Usize,
    Isize,
}

impl std::fmt::Display for NumSize {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            NumSize::F8 => write!(f, "f8"),
            NumSize::F16 => write!(f, "f16"),
            NumSize::F32 => write!(f, "f32"),
            NumSize::F64 => write!(f, "f64"),
            NumSize::I8 => write!(f, "i8"),
            NumSize::I16 => write!(f, "i16"),
            NumSize::I32 => write!(f, "i32"),
            NumSize::I64 => write!(f, "i64"),
            NumSize::U8 => write!(f, "u8"),
            NumSize::U16 => write!(f, "u16"),
            NumSize::U32 => write!(f, "u32"),
            NumSize::U64 => write!(f, "u64"),
            NumSize::Usize => write!(f, "usize"),
            NumSize::Isize => write!(f, "isize"),
        }
    }
}

impl TryFrom<&str> for NumSize {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "i8" => Ok(NumSize::I8),
            "i16" => Ok(NumSize::I16),
            "i32" => Ok(NumSize::I32),
            "i64" => Ok(NumSize::I64),
            "u8" => Ok(NumSize::U8),
            "u16" => Ok(NumSize::U16),
            "u32" => Ok(NumSize::U32),
            "u64" => Ok(NumSize::U64),
            "f8" => Ok(NumSize::F8),
            "f16" => Ok(NumSize::F16),
            "f32" => Ok(NumSize::F32),
            "f64" => Ok(NumSize::F64),
            "usize" => Ok(NumSize::Usize),
            "isize" => Ok(NumSize::Isize),
            _ => Err(()),
        }
    }
}

#[derive(Debug)]
pub enum Token<'tok> {
    Integer { value: u64, size: Option<NumSize> },
    SignedInteger { value: i64, size: Option<NumSize> },
    Float { value: f64, size: Option<NumSize> },
    Minus,
    String(&'tok str),
}

impl std::fmt::Display for Token<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Token::Integer { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::Float { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::SignedInteger { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::Minus => write!(f, "-"),
            Token::String(s) => write!(f, "{s}"),
        }
    }
}
