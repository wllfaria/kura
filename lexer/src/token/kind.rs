use miette::Error;

use super::{
    operator::Operator,
    token::{IntoToken, Token},
    value::Value,
};

#[derive(Debug)]
pub enum Kind<'tok> {
    Value(Value<'tok>),
    Op(Operator),

    Var,
    Const,
    Match,
    If,
    Else,
    Fun,
    Struct,
    Enum,
    Return,
    Eof,
}

impl<'tok> Kind<'tok> {
    pub fn infix_precedence(&self) -> Result<(u8, u8), Error> {
        match self {
            Kind::Op(Operator::Plus) | Kind::Op(Operator::Minus) => Ok((1, 2)),
            Kind::Op(Operator::Less)
            | Kind::Op(Operator::LessEqual)
            | Kind::Op(Operator::Greater)
            | Kind::Op(Operator::GreaterEqual) => Ok((3, 4)),
            Kind::Op(Operator::Star) | Kind::Op(Operator::Slash) => Ok((5, 6)),
            _ => miette::bail!("invalid operator with no precedence"),
        }
    }

    pub fn is_binary_op(&self) -> bool {
        matches!(
            self,
            Kind::Op(Operator::Star)
                | Kind::Op(Operator::Slash)
                | Kind::Op(Operator::Plus)
                | Kind::Op(Operator::Minus)
                | Kind::Op(Operator::Greater)
                | Kind::Op(Operator::Less)
                | Kind::Op(Operator::LessEqual)
                | Kind::Op(Operator::GreaterEqual)
        )
    }

    pub fn identifier_from(value: &'tok str) -> Kind<'tok> {
        match value {
            "var" => Kind::Var,
            "const" => Kind::Const,
            "match" => Kind::Match,
            "if" => Kind::If,
            "else" => Kind::Else,
            "fun" => Kind::Fun,
            "struct" => Kind::Struct,
            "enum" => Kind::Enum,
            "return" => Kind::Return,
            _ => Kind::Value(Value::Ident(value)),
        }
    }
}

impl<'tok> IntoToken<'tok> for Kind<'tok> {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok> {
        Token::new(self, (start_byte, end_byte).into())
    }
}
