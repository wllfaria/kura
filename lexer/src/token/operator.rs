use std::fmt;

use super::{
    kind::Kind,
    token::{IntoToken, Token},
};

#[derive(Debug)]
pub enum Operator {
    LeftParen,
    RightParen,
    LeftBracket,
    RightBracket,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    MinusEqual,
    Plus,
    PlusEqual,
    Equal,
    ThickArrow,
    EqualEqual,
    Star,
    StarEqual,
    Ampersand,
    Slash,
    SlashEqual,
    Colon,
    SemiColon,
    Bang,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    NotEqual,
    And,
    Or,
}

impl fmt::Display for Operator {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Operator::Minus => write!(f, "-"),
            Operator::MinusEqual => write!(f, "-="),
            Operator::Colon => write!(f, ":"),
            Operator::SemiColon => write!(f, ";"),
            Operator::LeftParen => write!(f, "("),
            Operator::RightParen => write!(f, ")"),
            Operator::LeftBracket => write!(f, "["),
            Operator::RightBracket => write!(f, "]"),
            Operator::LeftBrace => write!(f, "{{"),
            Operator::RightBrace => write!(f, "}}"),
            Operator::Comma => write!(f, ","),
            Operator::Dot => write!(f, "."),
            Operator::Plus => write!(f, "+"),
            Operator::PlusEqual => write!(f, "+="),
            Operator::Equal => write!(f, "="),
            Operator::ThickArrow => write!(f, "=>"),
            Operator::EqualEqual => write!(f, "=="),
            Operator::Star => write!(f, "*"),
            Operator::StarEqual => write!(f, "*="),
            Operator::Ampersand => write!(f, "&"),
            Operator::Slash => write!(f, "/"),
            Operator::Bang => write!(f, "!"),
            Operator::Less => write!(f, "<"),
            Operator::LessEqual => write!(f, "<="),
            Operator::Greater => write!(f, ">"),
            Operator::GreaterEqual => write!(f, ">="),
            Operator::NotEqual => write!(f, "!="),
            Operator::SlashEqual => write!(f, "/="),
            Operator::And => write!(f, "&&"),
            Operator::Or => write!(f, "||"),
        }
    }
}

impl<'tok> IntoToken<'tok> for Operator {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok> {
        Token::new(Kind::Op(self), (start_byte, end_byte).into())
    }
}
