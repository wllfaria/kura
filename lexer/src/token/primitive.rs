use core::fmt;

use super::{
    kind::Kind,
    token::{IntoToken, Token},
    value::Value,
};

#[derive(Debug)]
pub enum FloatSizes {
    F8,
    F16,
    F32,
    F64,
}

impl fmt::Display for FloatSizes {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            FloatSizes::F8 => write!(f, "f8"),
            FloatSizes::F16 => write!(f, "f16"),
            FloatSizes::F32 => write!(f, "f32"),
            FloatSizes::F64 => write!(f, "f64"),
        }
    }
}

#[derive(Debug)]
pub enum IntSizes {
    I8,
    I16,
    I32,
    I64,
    Isize,
}

impl fmt::Display for IntSizes {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            IntSizes::I8 => write!(f, "i8"),
            IntSizes::I16 => write!(f, "i16"),
            IntSizes::I32 => write!(f, "i32"),
            IntSizes::I64 => write!(f, "i64"),
            IntSizes::Isize => write!(f, "isize"),
        }
    }
}

#[derive(Debug)]
pub enum UIntSizes {
    U8,
    U16,
    U32,
    U64,
    Usize,
}

impl fmt::Display for UIntSizes {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            UIntSizes::U8 => write!(f, "u8"),
            UIntSizes::U16 => write!(f, "u16"),
            UIntSizes::U32 => write!(f, "u32"),
            UIntSizes::U64 => write!(f, "u64"),
            UIntSizes::Usize => write!(f, "usize"),
        }
    }
}

impl TryFrom<&str> for IntSizes {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "i8" => Ok(IntSizes::I8),
            "i16" => Ok(IntSizes::I16),
            "i32" => Ok(IntSizes::I32),
            "i64" => Ok(IntSizes::I64),
            "isize" => Ok(IntSizes::Isize),
            _ => Err(()),
        }
    }
}

impl TryFrom<&str> for FloatSizes {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "f8" => Ok(FloatSizes::F8),
            "f16" => Ok(FloatSizes::F16),
            "f32" => Ok(FloatSizes::F32),
            "f64" => Ok(FloatSizes::F64),
            _ => Err(()),
        }
    }
}

impl TryFrom<&str> for UIntSizes {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "u8" => Ok(UIntSizes::U8),
            "u16" => Ok(UIntSizes::U16),
            "u32" => Ok(UIntSizes::U32),
            "u64" => Ok(UIntSizes::U64),
            "usize" => Ok(UIntSizes::Usize),
            _ => Err(()),
        }
    }
}

#[derive(Debug)]
pub enum Primitive {
    Bool(bool),
    UInt {
        value: u64,
        size: Option<UIntSizes>,
    },
    Int {
        value: i64,
        size: Option<IntSizes>,
    },
    Float {
        value: f64,
        size: Option<FloatSizes>,
    },
}

impl fmt::Display for Primitive {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Primitive::Bool(b) => write!(f, "{b}"),
            Primitive::UInt { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Primitive::Float { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Primitive::Int { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
        }
    }
}

impl<'tok> IntoToken<'tok> for Primitive {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok> {
        Token::new(
            Kind::Value(Value::Primitive(self)),
            (start_byte, end_byte).into(),
        )
    }
}
