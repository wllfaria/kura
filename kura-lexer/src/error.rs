use crate::token::Location;

#[derive(Debug, Default)]
pub struct Error {
    location: Location,
    message: String,
}

impl Error {
    pub fn with_message(&mut self, message: String) {
        self.message = message;
    }
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} @ {}", self.message, self.location)
    }
}

impl std::error::Error for Error {}

impl From<std::ops::Range<usize>> for Error {
    fn from(value: std::ops::Range<usize>) -> Self {
        Self {
            location: value.into(),
            message: Default::default(),
        }
    }
}

impl From<Location> for Error {
    fn from(location: Location) -> Self {
        Self {
            location,
            message: Default::default(),
        }
    }
}
