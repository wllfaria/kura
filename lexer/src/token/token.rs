use std::{
    fmt,
    ops::{Range, RangeBounds},
};

use miette::SourceSpan;

use super::kind::Kind;

pub trait IntoToken<'tok> {
    fn into_token(self, start_byte: usize, end_byte: usize) -> Token<'tok>;
}

#[derive(Debug, Clone, Copy)]
pub struct Location {
    pub start_byte: usize,
    pub end_byte: usize,
}

impl Location {
    pub fn new(start: usize, end: usize) -> Self {
        Self {
            start_byte: start,
            end_byte: end,
        }
    }
}

impl From<Location> for SourceSpan {
    fn from(value: Location) -> Self {
        SourceSpan::new(value.start_byte.into(), value.end_byte - value.start_byte)
    }
}

impl From<(usize, usize)> for Location {
    fn from((start_byte, end_byte): (usize, usize)) -> Self {
        Self {
            start_byte,
            end_byte,
        }
    }
}

impl From<Range<usize>> for Location {
    fn from(range: Range<usize>) -> Self {
        let start = match range.start_bound() {
            std::ops::Bound::Included(start) => *start,
            std::ops::Bound::Excluded(start) => *start,
            std::ops::Bound::Unbounded => {
                panic!("can only construct a location from bounded ranges")
            }
        };

        let end = match range.end_bound() {
            std::ops::Bound::Included(end) => *end,
            std::ops::Bound::Excluded(end) => *end,
            std::ops::Bound::Unbounded => {
                panic!("can only construct a location from bounded ranges")
            }
        };

        Location {
            start_byte: start,
            end_byte: end,
        }
    }
}

#[derive(Debug)]
pub struct Token<'tok> {
    pub kind: Kind<'tok>,
    pub location: Location,
}

impl<'tok> Token<'tok> {
    pub fn new(kind: Kind<'tok>, location: Location) -> Self {
        Self { kind, location }
    }
}

impl fmt::Display for Token<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.kind)
    }
}
