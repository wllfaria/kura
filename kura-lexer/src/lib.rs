pub mod error;
pub mod token;

use error::Error;
use token::{FloatSizes, IntSizes, IntoToken, Kind, Operator, Primitive, Token, UIntSizes};

pub trait TransposeRef<'a, T, E: std::error::Error> {
    fn transpose(self) -> Result<Option<&'a T>, &'a E>;
}

impl<'lex> TransposeRef<'lex, Token<'lex>, Error> for Option<&'lex Result<Token<'lex>, Error>> {
    fn transpose(self) -> Result<Option<&'lex Token<'lex>>, &'lex Error> {
        match self {
            Some(result) => match result {
                Ok(token) => Ok(Some(token)),
                Err(e) => Err(e),
            },
            None => Ok(None),
        }
    }
}

pub struct Lexer<'lex> {
    pos: usize,
    source: &'lex str,
    pub complete_source: &'lex str,
    peeked: Option<Result<Token<'lex>, Error>>,
}

impl<'lex> Lexer<'lex> {
    pub fn new(source: &'lex str) -> Self {
        Self {
            pos: 0,
            source,
            complete_source: source,
            peeked: None,
        }
    }

    pub fn source_code(&self) -> &str {
        self.complete_source
    }

    pub fn peek(&mut self) -> Option<&Result<Token<'lex>, Error>> {
        if self.peeked.is_none() {
            self.peeked = self.next();
        }

        self.peeked.as_ref()
    }

    pub fn is_empty(&mut self) -> bool {
        self.peek().is_none()
    }

    fn make_token<T>(&mut self, tokenizable: T, size: usize) -> Token<'lex>
    where
        T: IntoToken<'lex>,
    {
        let start_byte = self.pos;
        self.advance_by(size);
        tokenizable.into_token(start_byte, self.pos)
    }

    pub fn expect(&mut self, expected: Kind<'_>) -> Result<Token<'lex>, Error> {
        let Some(token) = self.next().transpose()? else {
            let location = self.complete_source.len() - 1..self.complete_source.len();
            return Err(Error::from(location));
        };
        let kind = &token.kind;

        if kind == &expected {
            Ok(token)
        } else {
            Err(Error::from(token.location))
        }
    }

    pub fn expect_one_of(&mut self, expected_list: &[Kind<'_>]) -> Result<Token<'lex>, Error> {
        let token = self.next().transpose()?;
        let kind = token.as_ref().map(|token| &token.kind);
        let kind = kind.unwrap_or(&Kind::Eof);

        if expected_list.contains(kind) {
            Ok(token.unwrap())
        } else {
            let kinds = expected_list
                .iter()
                .map(|k| k.to_string())
                .collect::<Vec<_>>()
                .join(" ");
            panic!("invalid token kind. expected one of {kinds} but got {kind}");
        }
    }
}

impl<'lex> Iterator for Lexer<'lex> {
    type Item = Result<Token<'lex>, Error>;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            if let Some(peeked) = self.peeked.take() {
                return Some(peeked);
            }

            let mut chars = self.source.chars().peekable();
            let c = chars.next()?;

            let next = chars.peek().copied();

            break match (c, next) {
                (c, _) if c.is_whitespace() => {
                    self.advance_by(c.len_utf8());
                    continue;
                }
                // ----------------------------------------------------
                // DOUBLE TOKENS
                // ----------------------------------------------------
                ('+', Some('=')) => Some(Ok(self.make_token(Operator::PlusEqual, 2))),
                ('=', Some('=')) => Some(Ok(self.make_token(Operator::EqualEqual, 2))),
                ('=', Some('>')) => Some(Ok(self.make_token(Operator::ThickArrow, 2))),
                ('*', Some('=')) => Some(Ok(self.make_token(Operator::StarEqual, 2))),
                ('/', Some('=')) => Some(Ok(self.make_token(Operator::SlashEqual, 2))),
                ('!', Some('=')) => Some(Ok(self.make_token(Operator::NotEqual, 2))),
                ('<', Some('=')) => Some(Ok(self.make_token(Operator::LessEqual, 2))),
                ('>', Some('=')) => Some(Ok(self.make_token(Operator::GreaterEqual, 2))),
                ('-', Some('=')) => Some(Ok(self.make_token(Operator::MinusEqual, 2))),
                ('&', Some('&')) => Some(Ok(self.make_token(Operator::And, 2))),
                ('|', Some('|')) => Some(Ok(self.make_token(Operator::Or, 2))),
                ('-', Some(c)) if c.is_numeric() => Some(self.lex_numerals()),
                // we ignore a comment until the end of the line
                ('/', Some('/')) => {
                    let eol_location = self.source.find(|c| matches!(c, '\n')).unwrap_or(self.source.len());
                    self.advance_by(eol_location);
                    continue;
                }
                // for multiline comments, we ignore until we find the closing pattern
                ('/', Some('*')) => {
                    // we start past the next 2 characters as we know they are `/*`
                    let mut pos = 2;
                    let mut open_comments = 1;
                    loop {
                        if pos > self.source.len() - 1 {
                            break;
                        }

                        let curr = self.source.chars().nth(pos);
                        let next = self.source.chars().nth(pos + 1);

                        match (curr, next) {
                            (Some('*'), Some('/')) => open_comments -= 1,
                            (Some('/'), Some('*')) => open_comments += 1,
                            _ => (),
                        }

                        pos += 1;

                        if open_comments == 0 {
                            pos += 1;
                            break;
                        }
                    }

                    self.advance_by(pos);

                    continue;
                }
                // ----------------------------------------------------
                // SINGLE TOKENS
                // ----------------------------------------------------
                ('(', _) => Some(Ok(self.make_token(Operator::LeftParen, 1))),
                (')', _) => Some(Ok(self.make_token(Operator::RightParen, 1))),
                ('[', _) => Some(Ok(self.make_token(Operator::LeftBracket, 1))),
                (']', _) => Some(Ok(self.make_token(Operator::RightBracket, 1))),
                ('{', _) => Some(Ok(self.make_token(Operator::LeftBrace, 1))),
                ('}', _) => Some(Ok(self.make_token(Operator::RightBrace, 1))),
                (':', _) => Some(Ok(self.make_token(Operator::Colon, 1))),
                (';', _) => Some(Ok(self.make_token(Operator::SemiColon, 1))),
                (',', _) => Some(Ok(self.make_token(Operator::Comma, 1))),
                ('.', _) => Some(Ok(self.make_token(Operator::Dot, 1))),
                ('+', _) => Some(Ok(self.make_token(Operator::Plus, 1))),
                ('=', _) => Some(Ok(self.make_token(Operator::Equal, 1))),
                ('*', _) => Some(Ok(self.make_token(Operator::Star, 1))),
                ('&', _) => Some(Ok(self.make_token(Operator::Ampersand, 1))),
                ('/', _) => Some(Ok(self.make_token(Operator::Slash, 1))),
                ('!', _) => Some(Ok(self.make_token(Operator::Bang, 1))),
                ('<', _) => Some(Ok(self.make_token(Operator::Less, 1))),
                ('>', _) => Some(Ok(self.make_token(Operator::Greater, 1))),
                ('-', _) => Some(Ok(self.make_token(Operator::Minus, 1))),

                ('a'..='z' | 'A'..='Z' | '_', _) => Some(Ok(self.lex_identifier())),
                ('0'..='9', _) => Some(self.lex_numerals()),
                _ => Some(Ok(Kind::Eof.into_token(self.pos, self.pos))),
            };
        }
    }
}

impl<'lex> Lexer<'lex> {
    fn lex_identifier(&mut self) -> Token<'lex> {
        let next_whitespace = self
            .source
            .find(|c| !matches!(c, 'a'..='z' | 'A'..='Z' | '_' | '0'..='9'))
            .unwrap_or(self.source.len());

        let identifier = &self.source[..next_whitespace];

        self.make_token(Kind::identifier_from(identifier), next_whitespace)
    }

    fn lex_numerals(&mut self) -> Result<Token<'lex>, Error> {
        let start_byte = self.pos;
        let end_of_numeral = self
            .source
            .find(|c| !matches!(c, '-' | '_' | '.' | '0'..='9'))
            .unwrap_or(self.source.len());

        let mut literal = &self.source[..end_of_numeral];

        let mut dotted = literal.splitn(3, '.');
        match (dotted.next(), dotted.next(), dotted.next()) {
            (Some(one), Some(two), Some(_)) => {
                // numeric literal contains two or more dots, so we only lex
                // from the first numerical to the last numerical before the
                // second dot, <one.len()>.<two.len()>
                //
                // 123.456.789 -> 123.456 (leave .789) to be lexed after
                literal = &literal[..one.len() + 1 + two.len()];
            }
            _ => {
                // literal is either a integer, or a valid floating pointer
                // numeral, so we don't have to do anything here
            }
        }

        let mut dashed = literal.splitn(3, '-');
        match (dashed.next(), dashed.next(), dashed.next()) {
            (Some(""), Some(two), Some(_)) => {
                // numeric literal contain two or more dashes, but one of the
                // dashes is the first character, meaning we should only lex the
                // middle part as a signed integer
                //
                // -123.456-123 ("", "123.456", "123") (result of splitn)
                literal = &literal[..two.len() + 1]; // +1 to account for the `-`
            }
            (Some(one), Some(_), Some(_)) => {
                // numeric literal has two or more dashes, and don't start with a
                // dash, so we only parse the first part and ignore the rest.
                //
                // 123-456.78-9 ("123", "456.78", "9")
                literal = &literal[..one.len()];
            }
            _ => {
                // numeral is a unsigned integer, so we don't have to do
                // anything here
            }
        }

        // NOTE: we have to update the lexer state here before we potentially
        // change the size of the literal after trimming underscores
        let bytes_eaten = literal.len();
        self.pos += bytes_eaten;
        self.source = &self.source[bytes_eaten..];

        // we accept underscores in numerals, but rust parser don't, so we
        // trim here before parsing to rust's parser
        let literal = literal.replace('_', "");

        // numbers are allowed to potentially have a postfix specifying its size
        // like `1234usize`
        let end_of_postfix = self
            .source
            .find(|c: char| c.is_whitespace())
            .unwrap_or(self.source.len());

        let mut postfix = &self.source[..end_of_postfix];
        if postfix.ends_with(';') {
            postfix = postfix.trim_matches(';');
        }

        let is_signed = literal.contains('-');
        let is_float = literal.contains('.');
        let token = match (is_float, is_signed) {
            (false, true) => Primitive::Int {
                value: match literal.parse() {
                    Ok(numeral) => numeral,
                    Err(_) => {
                        return Err(Error::from(self.pos - bytes_eaten..self.pos));
                    }
                },
                size: IntSizes::try_from(postfix).ok(),
            },
            (false, false) => Primitive::UInt {
                value: match literal.parse() {
                    Ok(numeral) => numeral,
                    Err(_) => return Err(Error::from(self.pos - bytes_eaten..self.pos)),
                },
                size: UIntSizes::try_from(postfix).ok(),
            },
            (true, _) => Primitive::Float {
                value: match literal.parse() {
                    Ok(numeral) => numeral,
                    Err(_) => return Err(Error::from(self.pos - bytes_eaten..self.pos)),
                },
                size: FloatSizes::try_from(postfix).ok(),
            },
        };

        let has_size = match &token {
            Primitive::Int { size, .. } => size.is_some(),
            Primitive::UInt { size, .. } => size.is_some(),
            Primitive::Float { size, .. } => size.is_some(),
            _ => unreachable!(),
        };

        if has_size {
            self.source = &self.source[postfix.len()..];
            self.pos += postfix.len();
        }

        Ok(token.into_token(start_byte, self.pos))
    }

    fn advance_by(&mut self, amount: usize) {
        self.source = &self.source[amount..];
        self.pos += amount;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_sut(source: &str) -> Lexer<'_> {
        Lexer::new(source)
    }

    #[test]
    fn lexing_numericals() {
        let source = [
            // positive integers and floats
            "12346.67890",
            "1234567890",
            "9__876___5.43____210",
            "9_876_5.43_210",
            "9_____876__543____210",
            "9_876_543_210",
            // negative integers and floats
            "-9_____876__5.43____210",
            "-123_46.678_90",
            "-12346.67890",
            "-1234567890",
            "-9_____876__543____210",
            "-9_876_543_210",
            "-9_876_-5-43_210",
            // integers and floats with postfixes
            "-123456789i32",
            "123456789usize",
            "123.456789f64",
            "123.456789f64",
            "-12f8",
            "3.14159265358979323846264338327950288_f32",
            "3.141592653f32;",
        ];
        let source = source.join("\n");

        let mut numerals = vec![];

        for token in make_sut(&source) {
            numerals.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(numerals);
    }

    #[test]
    fn lexing_punctuations() {
        let source = [
            "()", "[]", "{}", ",", ".", "+", "-", "=", "*", "&", "*=", "+=", "-=", "/=", "!", "!=", "==", "<=", ">=",
            "<", ">", "/", ":", ";", "&&", "||",
        ];
        let source = source.join(" ");

        let mut punctuations = vec![];
        for token in make_sut(&source) {
            punctuations.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(punctuations);
    }

    #[test]
    fn lexing_builtin_identifiers() {
        let source = ["var", "const", "match", "if", "else", "fun", "struct", "enum", "return"];

        let source = source.join(" ");

        let mut builtin_identifiers = vec![];
        for token in make_sut(&source) {
            builtin_identifiers.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(builtin_identifiers);
    }

    #[test]
    fn lexing_simple_codeblock() {
        let source = [
            "fun calculate_circumference(diameter: f64) => f64 {",
            "   const pi = 3.14159265358979323846264338327950288_f32;",
            "   const radius = diameter / 2.0;",
            "   const circumference = 2.0 * pi * radius;",
            "",
            "   circumference",
            "}",
        ];

        let source = source.join("\n");

        let mut calculate_circumference_function = vec![];
        for token in make_sut(&source) {
            calculate_circumference_function.push(token.unwrap());
        }

        insta::assert_debug_snapshot!(calculate_circumference_function);
    }
}
