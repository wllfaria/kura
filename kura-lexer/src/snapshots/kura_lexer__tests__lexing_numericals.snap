---
source: kura-lexer/src/lib.rs
expression: numerals
---
[
    Token {
        kind: Value(
            Primitive(
                Float {
                    value: 12346.6789,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 0,
            end_byte: 11,
        },
    },
    Token {
        kind: Value(
            Primitive(
                UInt {
                    value: 1234567890,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 12,
            end_byte: 22,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Float {
                    value: 98765.4321,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 23,
            end_byte: 43,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Float {
                    value: 98765.4321,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 44,
            end_byte: 58,
        },
    },
    Token {
        kind: Value(
            Primitive(
                UInt {
                    value: 9876543210,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 59,
            end_byte: 80,
        },
    },
    Token {
        kind: Value(
            Primitive(
                UInt {
                    value: 9876543210,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 81,
            end_byte: 94,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Float {
                    value: -98765.4321,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 95,
            end_byte: 118,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Float {
                    value: -12346.6789,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 119,
            end_byte: 133,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Float {
                    value: -12346.6789,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 134,
            end_byte: 146,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Int {
                    value: -1234567890,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 147,
            end_byte: 158,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Int {
                    value: -9876543210,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 159,
            end_byte: 181,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Int {
                    value: -9876543210,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 182,
            end_byte: 196,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Int {
                    value: -9876,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 197,
            end_byte: 204,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Int {
                    value: -5,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 204,
            end_byte: 206,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Int {
                    value: -43210,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 206,
            end_byte: 213,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Int {
                    value: -123456789,
                    size: Some(
                        I32,
                    ),
                },
            ),
        ),
        location: Location {
            start_byte: 214,
            end_byte: 227,
        },
    },
    Token {
        kind: Value(
            Primitive(
                UInt {
                    value: 123456789,
                    size: Some(
                        Usize,
                    ),
                },
            ),
        ),
        location: Location {
            start_byte: 228,
            end_byte: 242,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Float {
                    value: 123.456789,
                    size: Some(
                        F64,
                    ),
                },
            ),
        ),
        location: Location {
            start_byte: 243,
            end_byte: 256,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Float {
                    value: 123.456789,
                    size: Some(
                        F64,
                    ),
                },
            ),
        ),
        location: Location {
            start_byte: 257,
            end_byte: 270,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Int {
                    value: -12,
                    size: None,
                },
            ),
        ),
        location: Location {
            start_byte: 271,
            end_byte: 274,
        },
    },
    Token {
        kind: Value(
            Ident(
                "f8",
            ),
        ),
        location: Location {
            start_byte: 274,
            end_byte: 276,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Float {
                    value: 3.141592653589793,
                    size: Some(
                        F32,
                    ),
                },
            ),
        ),
        location: Location {
            start_byte: 277,
            end_byte: 318,
        },
    },
    Token {
        kind: Value(
            Primitive(
                Float {
                    value: 3.141592653,
                    size: Some(
                        F32,
                    ),
                },
            ),
        ),
        location: Location {
            start_byte: 319,
            end_byte: 333,
        },
    },
    Token {
        kind: Op(
            SemiColon,
        ),
        location: Location {
            start_byte: 333,
            end_byte: 334,
        },
    },
]
